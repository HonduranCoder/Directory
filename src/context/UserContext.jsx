import { createContext, useContext, useState } from 'react';
import { getUser } from '../services/users';

const UserContext = createContext();
//Provider: provide state to other components
const UserProvider = ({ children }) => {
  const currentUser = getUser();
  const [user, setUser] = useState(
    //as the user in our state
    //expression that asks JS a question, JS will check and answer the question for us.
    currentUser ? { id: currentUser.id, email: currentUser.email } : null
  );
  return (
    //value is a React prop
    <UserContext.Provider value={{ user, setUser }}>
      {children}
    </UserContext.Provider>
  );
};
//custom hook to use throughout (function)
const useUser = () => {
  const userCtx = useContext(UserContext);
  if (userCtx === undefined) {
    throw new Error('Must be wrapped in a UserProvider to use useUser');
  }
  return userCtx;
};

export { UserProvider, useUser };
