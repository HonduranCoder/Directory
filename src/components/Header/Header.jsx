//Employee Directory
//signed in as: email
//sign out button
//A “Sign In” button if the user is signed out
//A “You are signed in as ${email}”` message with a “Sign Out” button if the user is signed in
import { useUser } from '../../context/UserContext.jsx';
import { Link } from 'react-router-dom';
import { signOutUser } from '../../services/users.js';
import { useHistory } from 'react-router-dom';

export default function Header() {
  const { user, setUser } = useUser();
  const history = useHistory();

  const handleSignOut = async () => {
    await signOutUser();
    setUser(null);
    history.push('/login');
  };

  return (
    <div>
      <span>Directory</span>{' '}
      <span>
        {user ? (
          <>
            You are signed in as {user.email}{' '}
            <button onClick={handleSignOut}> Sign out </button>{' '}
          </>
        ) : (
          <Link to="/login">
            <button>Sign in</button>
          </Link>
        )}
      </span>
    </div>
  );
}
//wrap in fragment because it needs a parent, they can't be siblings.
