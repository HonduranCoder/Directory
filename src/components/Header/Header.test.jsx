import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import Header from './Header.jsx';
import { UserProvider } from '../../context/UserContext.jsx';
//when mocking, look at the real component & find where you are calling code outside of your own application. (API call or backend call).

//follow the path, look for the mock inside mocks
//mock needs to match real path
//also telling it to look for a function inside of mock (a mock file function).
//mocking the functions that would hit the API
jest.mock('../../services/users.js');

//in the mock, there is no call to the backend.
//mockUser instead of getUser();
jest.mock('../../context/UserContext.jsx');

it('should render form', () => {
  const { container } = render(
    <UserProvider>
      <MemoryRouter>
        <Header />
      </MemoryRouter>
    </UserProvider>
  );
  expect(container).toMatchSnapshot();
});

it('should render form', () => {
  const { container } = render(
    <UserProvider mockUser={{ email: 's@s.com' }}>
      <MemoryRouter>
        <Header />
      </MemoryRouter>
    </UserProvider>
  );
  expect(container).toMatchSnapshot();
});
