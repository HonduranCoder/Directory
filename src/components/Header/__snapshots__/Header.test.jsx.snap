// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render form 1`] = `
<div>
  <div>
    <span>
      Directory
    </span>
     
    <span>
      <a
        href="/login"
      >
        <button>
          Sign in
        </button>
      </a>
    </span>
  </div>
</div>
`;

exports[`should render form 2`] = `
<div>
  <div>
    <span>
      Directory
    </span>
     
    <span>
      You are signed in as 
      s@s.com
       
      <button>
         Sign out 
      </button>
       
    </span>
  </div>
</div>
`;
